README.md
# NAME

Test::Mojo::Trim - Trim strings for Test::Mojo

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-test-mojo-trim"><img src="https://api.travis-ci.org/Csson/p5-test-mojo-trim.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Test-Mojo-Trim-0.1000"><img src="https://badgedepot.code301.com/badge/kwalitee/Test-Mojo-Trim/0.1000" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Test-Mojo-Trim%200.1000"><img src="https://badgedepot.code301.com/badge/cpantesters/Test-Mojo-Trim/0.1000" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-68.4%-red.svg" alt="coverage 68.4%" />
    </p>
</div>

# VERSION

Version 0.1000, released 2016-07-22.

# SYNOPSIS

    use Mojo::Base -strict;
    use Mojolicious::Lite;
    use Test::More;
    use Test::Mojo::Trim;

    my $test = Test::Mojo::Trim->new;

    get '/test_1';

    my $compared_to = qq{ <div><h1>Header</h1><p>A paragraph.</p></div> };

    $test->get_ok('/test_1')->status_is(200)->trimmed_content_is($compared_to);

    done_testing();

    __DATA__
    @@ the_test.html.ep
    <div>
        <h1>Header</h1>
        <p>A paragraph.</p>
    </div>

# DESCRIPTION

Test::Mojo::Trim is an extension to Test::Mojo, that adds an additional string comparison function.

# METHODS

[Test::Mojo::Trim](https://metacpan.org/pod/Test::Mojo::Trim) inherits all methods from [Test::Mojo](https://metacpan.org/pod/Test::Mojo) and implements the following new one.

## trimmed\_content\_is

    $test->get_ok('/test')->trimmed_content_is('<html></html>');

Removes all whitespace between tags from the two strings that are compared.
That is, if a > and < is separated only by whitespace, that whitespace is removed.
Any leading or trailing whitespace is also removed.

# SEE ALSO

- [Test::Mojo](https://metacpan.org/pod/Test::Mojo)
- [Test::Mojo::Most](https://metacpan.org/pod/Test::Mojo::Most)

# SOURCE

[https://github.com/Csson/p5-test-mojo-trim](https://github.com/Csson/p5-test-mojo-trim)

# HOMEPAGE

[https://metacpan.org/release/Test-Mojo-Trim](https://metacpan.org/release/Test-Mojo-Trim)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Test-Mojo-Trim

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.1000 2016-07-22T12:46:21Z UTC
 [Bug Fixes]
 - Fixed Mojolicious 7.0 compatibility (fixes #2).

 [Meta]
 - Updated iller.yaml to current syntax.

 [Stats & Dependency Changes]
 - 5 files changed, 50 insertions(+), 28 deletions(-)
 - (run req) ~ perl 5.010001 --> 5.010000
 - (test req) + ExtUtils::MakeMaker (any)
 - (test req) + File::Spec (any)
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (test rec) + CPAN::Meta 2.120900
 - (conf req) + ExtUtils::MakeMaker (any)
 - (conf req) - Module::Build::Tiny
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::ReportPrereqs 0.021
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::ContributorsFromGit
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::ModuleBuildTiny
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0310

 [Testing]
 - Added a real test.

0.0904 2015-02-12T21:55:01Z CET
 [Requirements]
 - Removed an unused Pod::Weaver plugin from prereqs.

0.0903 2015-02-12T18:18:19Z CET
 [Enhancements]
 - Updated usage of deprecated syntax to current api.

 [Meta]
 - Removed cpanfile, now prereqs are set in dist.ini.

0.0902 2015-02-06T23:23:26Z CET
 [Revision]
 - Bump version number to pass tests.

0.0901 2015-02-06T23:16:29Z CET
 [Documentation]
 - Rebuilt with Podweaver.

 [Meta]
 - No functional changes.
 - Changed versioning to x.yyyy.
 - Now uses @Author::CSSON dist zilla bundle, expanded.
 - Removed the last traces of Dist::Milla.

0.0900 2014-12-04T15:30:57Z CET
 - Better use of Mojo::DOM api.

0.08 2014-09-18T21:19:44Z CEST
 - In case of request failer (ie status 500) the result is parsed and
   output in a more readable style.
 - Fix missing build prereqs.

0.07 2014-09-06T15:31:51Z CEST
 - Added Travis build status.
 - Added [Test::EOL] and [Test::EOF]

0.06 2014-09-06T07:29:03Z CEST
 - Fixed faulty Perl version requirement

0.05 2014-09-04T20:44:07Z CEST
 - Fixed a cosmetic problem related to last release where it would think
   that a request error had occured

0.04 2014-09-04T16:44:56Z CEST
 - If the request fails it dumps the Mojolicious error page, which is hard
   ro read. This release extracts the specific action and adds it to the
   test description.

0.03 2014-08-31T18:06:32Z CEST
 - Improved test message

0.02 2014-08-31T11:23:53Z CEST
 - Added missing dependency

0.01 2014-08-31T10:46:14Z CEST
 - Initial release.


lib/Test/Mojo/Trim.pm
package Test::Mojo::Trim;

use strict;

use Mojo::DOM;
use Mojo::Base 'Test::Mojo';
use Mojo::Util 'trim';

# ABSTRACT: Trim strings for Test::Mojo
# AUTHORITY
our $VERSION = '0.1001';

sub trimmed_content_is {
    my $self = shift;
    my $value = squish(Mojo::DOM->new(shift)->to_string);
    my $desc = shift;

    my $dom = $self->tx->res->dom;
    my $got = squish($dom->to_string);
    my $error = defined $dom->at('#error') ? $dom->at('#error')->text : undef;
    chomp $error if $error;

    $value =~ s{> <}{><}g;
    $got =~ s{> <}{><}g;
    $desc ||= 'exact match for trimmed content';

    if(defined $error && length $error) {
        $desc .= (defined $error && length $error ? " (Error: $error)" : '');
        my $table = $dom->find('#context table')->each(sub {
            $desc .= $_->find('td')->map(sub { $_->text })->join(' ');
        });
        $got = '<see error>';
    }

    return $self->_test('is', $got, $value, $desc);
}

sub squish {
    my $string = trim @_;
    $string =~ s{\s+}{ }g;
    return $string;
}

1;
__END__

=encoding utf-8


=head1 SYNOPSIS

    use Mojo::Base -strict;
    use Mojolicious::Lite;
    use Test::More;
    use Test::Mojo::Trim;

    my $test = Test::Mojo::Trim->new;

    get '/test_1';

    my $compared_to = qq{ <div><h1>Header</h1><p>A paragraph.</p></div> };

    $test->get_ok('/test_1')->status_is(200)->trimmed_content_is($compared_to);

    done_testing();

    __DATA__
    @@ the_test.html.ep
    <div>
        <h1>Header</h1>
        <p>A paragraph.</p>
    </div>

=head1 DESCRIPTION

Test::Mojo::Trim is an extension to Test::Mojo, that adds an additional string comparison function.

=head1 METHODS

L<Test::Mojo::Trim> inherits all methods from L<Test::Mojo> and implements the following new one.

=head2 trimmed_content_is

    $test->get_ok('/test')->trimmed_content_is('<html></html>');

Removes all whitespace between tags from the two strings that are compared.
That is, if a E<gt> and E<lt> is separated only by whitespace, that whitespace is removed.
Any leading or trailing whitespace is also removed.

=head1 SEE ALSO

=for :list
* L<Test::Mojo>
* L<Test::Mojo::Most>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.006.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Trim strings for Test::Mojo",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-Mojo-Trim",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Test::Mojo::Trim",
  "PREREQ_PM" => {
    "Mojolicious" => "5.00"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.1001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Mojolicious" => "5.00",
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
